File: Python/Árvore Binária/ArvoreBinaria.py
class No:
    def __init__(self, conteudo):
        self.conteudo = conteudo
        self.noEsq = None
        self.noDir = None


class ArvoreBinaria:
    def __init__(self) -> None:
        self.raiz = None


    # INSERT
    def Insert(self, conteudo):
        novoNo = No(conteudo)
        self.raiz = self.insert(self.raiz, novoNo)

    def insert(self, atual, novoNo):
        if atual == None:
            return novoNo;
        elif atual.conteudo > novoNo.conteudo:
            atual.noEsq = self.insert(atual.noEsq, novoNo)
        else:
            atual.noDir = self.insert(atual.noDir, novoNo)

        return atual


    # REMOVE
    def remove(self, valor, atual=None):
        if atual is None:
            atual = self.raiz

        if valor < atual.conteudo:
            atual.noEsq = self.remove(valor, atual.noEsq)
        elif valor > atual.conteudo:
            atual.noDir = self.remove(valor, atual.noDir)

        else:
            if atual.noEsq is None:
                aux = atual.noDir
                atual = None
                return aux
            
            elif atual.noDir is None:
                aux = atual.noEsq
                atual = None
                return aux
            
            else:
                aux = self.menorDireita(atual.noDir)
                atual.conteudo = aux.conteudo
                atual.noDir = self.remove(aux.conteudo, atual.noDir)

        return atual

    def menorDireita(self, no):
        while no.noEsq:
            no = no.noEsq
        return no


    # INORDER
    def InOrder(self):
        print("Exibição inOrder: ")
        self.inOrder(self.raiz)
        print("None\n")

    def inOrder(self, atual):
        if atual != None:
            self.inOrder(atual.noEsq)
            print(atual.conteudo, end=" -> ")
            self.inOrder(atual.noDir)


    # PREORDER
    def PreOrder(self):
        print("Exibição PreOrder: ")
        self.preOrder(self.raiz)
        print("None\n")

    def preOrder(self, atual):
        if atual != None:
            print(atual.conteudo, end=" -> ")
            self.preOrder(atual.noEsq)
            self.preOrder(atual.noDir)


    # POSORDER
    def PosOrder(self):
        print("Exibição PosOrder: ")
        self.posOrder(self.raiz)
        print("None\n")

    def posOrder(self, atual):
        if atual != None:
            self.posOrder(atual.noEsq)
            self.posOrder(atual.noDir)
            print(atual.conteudo, end=" -> ")

File: JavaScript/Árvore BInária/ArvoreBinaria.js
class No {
    constructor(conteudo) {
        this.noEsq = null
        this.noDir = null
        this.conteudo = conteudo
    }
}


class ArvoreBinaria {
    constructor () {
        this.raiz = null;
    }


    insert(conteudo) {
        const novoNo = new No(conteudo);
        this.raiz = this.Insert(this.raiz, novoNo)
    }

    Insert(noAtual, novoNo) {
        if (noAtual == null) {
            return novoNo;

        } else if (novoNo.conteudo > noAtual.conteudo) {
            noAtual.noDir = this.Insert(noAtual.noDir, novoNo);

        } else if (novoNo.conteudo < noAtual.conteudo) {
            noAtual.noEsq = this.Insert(noAtual.noEsq, novoNo);
        }

        return noAtual;
    }


    inOrder() {
        console.log("Exibição InOrder:");
        this.InOrder(this.raiz);
    }

    InOrder(noAtual) {
        if (noAtual != null) {
            this.InOrder(noAtual.noEsq);
            console.log(noAtual.conteudo);
            this.InOrder(noAtual.noDir);
        }
    }

    preOrder() {
        console.log("Exibição PreOrder:");
        this.PreOrder(this.raiz);
    }

    PreOrder(noAtual) {
        if (noAtual != null) {
            console.log(noAtual.conteudo);
            this.PreOrder(noAtual.noEsq);
            this.PreOrder(noAtual.noDir);
        }
    }

    posOrder() {
        console.log("Exibição PosOrder:");
        this.PosOrder(this.raiz);
    }

    PosOrder(noAtual) {
        if (noAtual != null) {
            this.PosOrder(noAtual.noEsq);
            this.PosOrder(noAtual.noDir);
            console.log(noAtual.conteudo);
        }
    }


    remove(valor, noAtual = null) {
        if (noAtual == null) {
            noAtual = this.raiz;
        }

        if (valor > noAtual.conteudo) {
            noAtual.noDir = this.remove(valor, noAtual.noDir);

        } else if (valor < noAtual.conteudo) {
            noAtual.noEsq = this.remove(valor, noAtual.noEsq);

        } else {
            if (noAtual.noEsq == null) {
                let aux = noAtual.noDir;
                noAtual = null;
                return aux;

            } else if (noAtual.noDir == null) {
                let aux = noAtual.noEsq;
                noAtual = null;
                return aux;

            } else {
                let aux = this.menorDireita(noAtual.noDir);
                noAtual.conteudo = aux.conteudo;
                noAtual.noDir = this.remove(aux.conteudo, noAtual.noDir);
            }
        }

        return noAtual;
    }

    menorDireita(noAtual) {
        while (noAtual.noEsq) {
            noAtual = noAtual.noEsq;
        }

        return noAtual;
    }
}

const arvoreBinaria = new ArvoreBinaria();

arvoreBinaria.insert(5)
arvoreBinaria.insert(8)
arvoreBinaria.insert(2)
arvoreBinaria.insert(4)
arvoreBinaria.insert(9)

console.log(arvoreBinaria.inOrder())
console.log(arvoreBinaria.preOrder())
console.log(arvoreBinaria.posOrder())

arvoreBinaria.remove(9)

console.log(arvoreBinaria.inOrder())
console.log(arvoreBinaria.preOrder())
console.log(arvoreBinaria.posOrder())